### PR TITLE
[feat]: [CDS-42743]: Azure Artifacts Connector

### PR DESCRIPTION
**Pull request by vtxorxwitty**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/37117)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 5 Issues 
* Code issue 5 Code issues 

 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/37117)

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/azureartifacts/AzureArtifactsConnectionTypeDTO.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ */
+
+package io.harness.delegate.beans.connector.azureartifacts;
+
+import io.harness.delegate.beans.connector.scm.GitConfigConstants;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public enum AzureArtifactsConnectionTypeDTO {
+  @JsonProperty(GitConfigConstants.PROJECT) PROJECT,
```
Comment (puthrayaharness) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/azureartifacts/AzureArtifactsConnectionTypeDTO.java:15]: Why are these GitConfigConstants?

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/azureartifacts/AzureArtifactsConnectionTypeDTO.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  @JsonProperty(GitConfigConstants.REPO) REPO
```
Comment (puthrayaharness) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/azureartifacts/AzureArtifactsConnectionTypeDTO.java:16]: Where are using these?

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/azureartifacts/AzureArtifactsConnectorConstants.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+package io.harness.delegate.beans.connector.azureartifacts;
+
+import io.harness.annotations.dev.HarnessTeam;
+import io.harness.annotations.dev.OwnedBy;
+
+@OwnedBy(HarnessTeam.CDC)
+public class AzureArtifactsConnectorConstants {
+  public static final String TOKEN = "Token";
+
+  public static final String USERNAME_AND_TOKEN = "UsernameToken";
```
Comment (puthrayaharness) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/azureartifacts/AzureArtifactsConnectorConstants.java:17]: Does Azure Artifacts support Username & Token?

File: 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java [hunk 4]
```diff
@@ -353,6 +353,7 @@ public enum FeatureName {
   TEMPLATE_SCHEMA_VALIDATION,
   YAML_APIS_GRANULAR_PERMISSION,
   JENKINS_BUILD,
+  NG_AZURE_ARTIFACTS,
```
Comment (puthrayaharness) [inline 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java:356]: `AZURE_ARTIFACTS_NG`

File: 960-api-services/src/main/java/io/harness/artifacts/azureartifacts/client/AzureArtifactsRestClientFactoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    } catch (InvalidArtifactServerException e) {
+      return azureDevopsUrl;
+    }
+  }
+
+  public static void validateAzureDevopsUrl(String azureDevopsUrl) {
+    try {
+      URI uri = new URI(azureDevopsUrl);
+    } catch (URISyntaxException e) {
+      throw new InvalidArtifactServerException("Azure DevOps URL is invalid");
```
Comment (puthrayaharness) [inline 960-api-services/src/main/java/io/harness/artifacts/azureartifacts/client/AzureArtifactsRestClientFactoryImpl.java:76]: Please throw Hint/Explanation exception

File: 960-api-services/src/main/java/io/harness/artifacts/azureartifacts/client/AzureArtifactsRegistryRestClient.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import io.harness.artifacts.azureartifacts.beans.*;
+
+import retrofit2.Call;
+import retrofit2.http.GET;
+import retrofit2.http.Header;
+import retrofit2.http.Path;
+import retrofit2.http.Query;
+
+@OwnedBy(CDC)
+public interface AzureArtifactsRegistryRestClient {
```
Comment (puthrayaharness) [inline 960-api-services/src/main/java/io/harness/artifacts/azureartifacts/client/AzureArtifactsRegistryRestClient.java:22]: Why are we recreating these? We already have all this in CG? We should be resuing the code from CG...

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

File: 440-connector-nextgen/src/main/java/io/harness/connector/mappers/azureartifacts/AzureArtifactsDTOToEntity.java [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+  }
+
+  private AzureArtifactsHttpAuth getHttpAuth(
+      AzureArtifactsHttpAuthenticationType type, AzureArtifactsHttpCredentialsDTO httpCredentialsDTO) {
+    if (type == null) {
+      throw new InvalidRequestException("AzureArtifacts Http Auth Type not found");
+    }
+
+    switch (type) {
+      case USERNAME_AND_TOKEN:
```
Comment (puthrayaharness) [inline 440-connector-nextgen/src/main/java/io/harness/connector/mappers/azureartifacts/AzureArtifactsDTOToEntity.java:120]: Where do we have username/Password? We just have token auth in Azure Artifacts?

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/azureartifacts/AzureArtifactsConnectorDTO.java [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+  @Valid
+  @NotNull
+  @Schema(description = "Details for authentication mechanism for AzureArtifacts connector")
+  AzureArtifactsAuthenticationDTO authentication;
+
+  @Valid @Schema(description = "API access details") AzureArtifactsApiAccessDTO apiAccess;
+
+  @Schema(description = "Selected Connectivity Modes") Set<String> delegateSelectors;
+
+  Boolean executeOnDelegate = true;
```
Comment (puthrayaharness) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/azureartifacts/AzureArtifactsConnectorDTO.java:77]: Why do need this? Where do we have executeOnDelegate?

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/azureartifacts/AzureArtifactsConnectorDTO.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+@OwnedBy(HarnessTeam.CDC)
+@Schema(name = "AzureArtifactsConfig", description = "This contains details of AzureArtifacts connector")
+public class AzureArtifactsConnectorDTO
+    extends ConnectorConfigDTO implements ScmConnector, DelegateSelectable, ManagerExecutable {
+  @NotNull
+  @JsonProperty("type")
+  @Schema(description = "Project or Repository connector type")
+  AzureArtifactsConnectionTypeDTO connectionType;
+
+  @NotBlank @NotNull @Schema(description = "SSH or HTTP URL based on type of connection") String url;
```
Comment (puthrayaharness) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/azureartifacts/AzureArtifactsConnectorDTO.java:63]: Where is SSH URL?

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/azureartifacts/AzureArtifactsConnectorDTO.java [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+      return url;
+    }
+    if (GitAuthType.SSH.equals(authentication.getAuthType())) {
+      return FilePathUtils.addEndingSlashIfMissing(url) + gitRepositoryDTO.getName();
+    }
+    return FilePathUtils.addEndingSlashIfMissing(url) + AZURE_ARTIFACTS_NAME_SEPARATOR + gitRepositoryDTO.getName();
+  }
+
+  @Override
+  public GitRepositoryDTO getGitRepositoryDetails() {
```
Comment (puthrayaharness) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/azureartifacts/AzureArtifactsConnectorDTO.java:149]: Where do we have GitRepo with Azure.

File: 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/azureartifacts/AzureArtifactsToGitMapper.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    final GitConnectionType connectionType = mapToGitConnectionType(connectorDTO.getConnectionType());
+    final String url = connectorDTO.getUrl();
+    final String validationRepo = connectorDTO.getValidationRepo();
+    if (authType == null) {
+      throw new InvalidRequestException("Azure Artifacts DTO Auth Type not found");
+    }
+    switch (authType) {
+      case HTTP:
+        return mapToGitHTTP(connectorDTO, connectionType, url, validationRepo);
+      case SSH:
```
Comment (puthrayaharness) [inline 954-connector-beans/src/main/java/io/harness/delegate/beans/connector/azureartifacts/AzureArtifactsToGitMapper.java:37]: Where do we have SSH?

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/azureartifacts/AzureArtifactsTaskHelper.java [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+          saveLogs(executionLogCallback, "Validating  Artifact Server");
+          artifactTaskResponse =
+              getSuccessTaskResponse(azureArtifactsTaskHandler.getAzureArtifactsProjects(attributes));
+          saveLogs(executionLogCallback, "validated artifact server: " + registryUrl);
+
+          break;
+
+        case GET_AZURE_FEEDS:
+
+          saveLogs(executionLogCallback, "Validating  Artifact Server");
```
Comment (puthrayaharness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/azureartifacts/AzureArtifactsTaskHelper.java:85]: Please fix the logs in the whole file

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

File: 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactResponseToOutcomeMapper.java [hunk 13]
```diff
@@ -143,12 +146,33 @@ public ArtifactOutcome toArtifactOutcome(
...
+            (AzureArtifactsDelegateResponse) artifactDelegateResponse;
+        return getAzureArtifactsOutcome(azureArtifactsConfig, azureArtifactsDelegateResponse, useDelegateResponse);
       default:
         throw new UnsupportedOperationException(
             String.format("Unknown Artifact Config type: [%s]", artifactConfig.getSourceType()));
     }
   }
 
+  private static AzureArtifactsOutcome getAzureArtifactsOutcome(AzureArtifactsConfig azureArtifactsConfig,
+      AzureArtifactsDelegateResponse azureArtifactsDelegateResponse, boolean useDelegateResponse) {
```
[the remark this answers is not in the document]
Comment (vtxorxwitty) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/artifact/mappers/ArtifactResponseToOutcomeMapper.java:168]: We are not using it anywhere. But this field is added for all the artifact source including docker, artifactory, s3, github...

**Pull request closed without merging**